File: tools/deploy_selected_hosting_files.py
import gzip,hashlib,json,subprocess,sys,requests
site='decorate-me'; base='https://firebasehosting.googleapis.com/v1beta1'
token=subprocess.check_output(['gcloud.cmd','auth','print-access-token'],text=True).strip()
h={'Authorization':f'Bearer {token}','x-goog-user-project':'decorate-me'}
def call(method,url,**kw):
 r=requests.request(method,base+url,headers=h,timeout=90,**kw); r.raise_for_status(); return r.json() if r.content else {}
releases=call('GET',f'/sites/{site}/releases?pageSize=1')['releases']; current=releases[0]['version']; source=current['name']
version=call('POST',f'/projects/-/sites/{site}/versions',json={'config':current['config'],'labels':{'deployment-tool':'codex-selected-files'}})['name']
files={}; page=''
while True:
 data=call('GET',f'/{source}/files?pageSize=1000'+(f'&pageToken={page}' if page else ''))
 for f in data.get('files',[]):
  if not (f['path'].startswith('/.git/') or f['path'].startswith('/.claude/')): files[f['path']]=f['hash']
 page=data.get('nextPageToken','')
 if not page: break
bodies={}
for arg in sys.argv[1:]:
 url,path=arg.split('=',1); body=gzip.compress(open(path,'rb').read(),compresslevel=9,mtime=0); digest=hashlib.sha256(body).hexdigest(); files[url]=digest;bodies[digest]=body
upload=''
entries=list(files.items())
for i in range(0,len(entries),1000):
 data=call('POST',f'/{version}:populateFiles',json={'files':dict(entries[i:i+1000])});upload=data['uploadUrl']
 for digest in data.get('uploadRequiredHashes',[]):
  if digest not in bodies: raise RuntimeError('unexpected missing old hash '+digest)
  r=requests.post(upload+'/'+digest,headers=h,data=bodies[digest],timeout=90);r.raise_for_status()
call('PATCH',f'/{version}?updateMask=status',json={'status':'FINALIZED'})
release=call('POST',f'/projects/-/sites/{site}/channels/live/releases?versionName={version}',json={'type':'DEPLOY','message':'Selected OTP, guest feedback, admin macro fixes; remove internal dotfiles'})
print(json.dumps({'version':version,'release':release.get('name'),'files':list(sys.argv[1:])}))

File: tools/deploy_selected_hosting_files.js
// Rebuild current Hosting manifest, remove private dot paths, replace only listed files, release.
const fs=require('fs'),zlib=require('zlib'),crypto=require('crypto'),path=require('path');
const api=require('firebase-tools/lib/hosting/api');
const {Client}=require('firebase-tools/lib/apiv2');
const site='decorate-me';
const files=Object.fromEntries(process.argv.slice(2).map(arg=>{const [url,...rest]=arg.split('=');return [url,rest.join('=')];}));
if(!Object.keys(files).length) throw Error('no files');
(async()=>{
  const client=new Client({urlPrefix:'https://firebasehosting.googleapis.com/v1beta1',auth:true});
  const releasePage=await client.get(`/sites/${site}/releases`,{queryParams:{pageSize:1}});
  const releases=releasePage.body.releases||[];
  const source=releases[0]?.version?.name;
  if(!source) throw Error('current version unavailable');
  const version=await api.createVersion(site,{config:releases[0].version.config,labels:{'deployment-tool':'codex-selected-files'}});
  const map={}; const bodies={};
  let pageToken;
  do {
    const queryParams={pageSize:1000}; if(pageToken) queryParams.pageToken=pageToken;
    const page=await client.get(`/${source}/files`,{queryParams});
    for(const f of page.body.files||[]){
      if(f.path.startsWith('/.git/')||f.path.startsWith('/.claude/')) continue;
      map[f.path]=f.hash;
    }
    pageToken=page.body.nextPageToken;
  } while(pageToken);
  for(const [url,file] of Object.entries(files)){
    const raw=fs.readFileSync(path.resolve(file));
    const gzip=zlib.gzipSync(raw,{level:9});
    const hash=crypto.createHash('sha256').update(gzip).digest('hex');
    map[url]=hash;bodies[hash]=gzip;
  }
  let uploadClient;
  for(let i=0;i<Object.keys(map).length;i+=1000){
    const batch=Object.fromEntries(Object.entries(map).slice(i,i+1000));
    const populated=await client.post(`/${version}:populateFiles`,{files:batch});
    uploadClient ||= new Client({urlPrefix:populated.body.uploadUrl,auth:true});
    for(const hash of populated.body.uploadRequiredHashes||[]){
      if(!bodies[hash]) throw Error('unexpected required hash '+hash);
      const res=await uploadClient.request({method:'POST',path:`/${hash}`,body:bodies[hash],resolveOnHTTPError:true});
      if(res.status!==200) throw Error('upload failed '+hash+' '+res.status);
    }
  }
  await api.updateVersion(site,version.split('/').pop(),{status:'FINALIZED'});
  const release=await api.createRelease(site,'live',version,{type:'DEPLOY',message:'Selected admin macro and OTP fixes; internal dotfiles removed from cloned release'});
  console.log(JSON.stringify({source,version,release:release.name,files:Object.keys(files),removed:['/.git/**','/.claude/**']}));
})().catch(e=>{console.error(e);process.exit(1)});
